[FLINK-25846][FLINK-25848] Async Sink does not gracefully shutdown on Cancel, KDS Sink does not fast fail when invalid configuration supplied

CrynetLogistics opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Bugfixes:
  Async Sink does not gracefully shutdown on Cancel 
  KDS Sink does not fast fail when invalid configuration supplied


## Brief change log

* Added more cases to the `isRetryable` method in the KDSSinkWriter to handle permanently failing cases.


## Verifying this change

* Added unit test cases

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)n
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)y
  - The serializers: (yes / no / don't know)n
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)n
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)n
  - The S3 file system connector: (yes / no / don't know)n

## Documentation

  - Does this pull request introduce a new feature? (yes / no)n
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) docs/javadocs

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsException.java:
```diff
@@ -39,12 +39,12 @@ public KinesisDataStreamsException(final String message, final Throwable cause)
 
         public KinesisDataStreamsFailFastException() {
             super(
-                    "Encountered an exception while persisting records, not retrying due to {failOnError} being set.");
```
Instead of duplicating this string, either call `this(null)` or extract to constant

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsException.java:
```diff
@@ -39,12 +39,12 @@ public KinesisDataStreamsException(final String message, final Throwable cause)
         public KinesisDataStreamsFailFastException(final Throwable cause) {
             super(
-                    "Encountered an exception while persisting records, not retrying due to {failOnError} being set.",
+                    "Encountered an exception while persisting records, not retrying due to either: {failOnError} being set or the exception should not be retried.",
```
Semantically it would make more sense to split the exception rather than adding ambiguity for user. For example, we could add a `KinesisDataStreamsFatalException`

dannycranmer replied:
On second thoughts is this even correct? Seems like we are throwing `KinesisDataStreamsException` in this PR

CrynetLogistics replied:
I'm currently wrapping any exception that I think is not retryable with the `KinesisDataStreamsException`, if we encounter it in the KDSSink.

I hear you, maybe it will be better to throw any `Error`s and `RuntimeException`s  directly without wrapping. This way these non-checked exceptions would be thrown without a KDSException wrapper confusing things - but obvs they must still be accepted by the consumer because even non-checked exceptions aren't fatal until they are consumed by the mailbox.

Also agree we can split failOnError and not retrying exceptions.

CrynetLogistics replied:
Ah nevermind, I understand now.

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -176,11 +179,35 @@ private void handlePartiallyFailedRequest(
+            getFatalExceptionCons().accept(new FlinkException("Running job was cancelled"));
+            return false;
+        }
+        if (err instanceof CompletionException
```
This code is complex, and hard to read. Suggest we use a helper, would this work?:

```
if (ExceptionUtils.findThrowable(ex, ResourceNotFoundException.class).isPresent()) {
     ....
}

```

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -176,11 +179,35 @@ private void handlePartiallyFailedRequest(
+                                    err));
+            return false;
+        }
+        if (err instanceof Error) {
+            getFatalExceptionCons()
```
Happy to merge to fix the bug, but this is too much logic in the implementation. Since this is common to all Sinks can we pull to parent?

vahmed-hamdy replied:
The parent (`AsynSinkWriter`) doesn't offer retry strategy leaving the request submission implementation to concrete classes. Is it better to enforce having a retry strategy on all implementations or just move the common logic to a validator/classifier class to be reused by all implementations without changing base class?

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -176,11 +179,35 @@ private void handlePartiallyFailedRequest(
+            getFatalExceptionCons()
+                    .accept(
+                            new KinesisDataStreamsException(
+                                    "Encountered non-recoverable exception relating to not being able to find the specified resources",
```
Is this message correct for all sub classes of `Error`? I think we should make this more generic, something like `Encountered non-recoverable exception`

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -56,6 +62,61 @@
 
+    private static final RetryValidationStrategy INVALID_CREDENTIALS_STRATEGY =
+            new RetryValidationStrategy(
+                    err -> ExceptionUtils.findThrowable(err, StsException.class).isPresent(),
```
nit: This is much better, but still duplicating logic. You could sub `RetryValidationStrategy` with a `RootCauseRetryValidationStrategy` that wraps up the `ExceptionUtils.findThrowable(` and you simply pass the class to look for

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -56,6 +62,61 @@
+    private static final RetryValidationStrategy MISSING_ACCESS_KEY_ID_STRATEGY =
+            new RetryValidationStrategy(
+                    err ->
+                            ExceptionUtils.findThrowableWithMessage(
+                                                    err, "Access key ID cannot be blank.")
+                                            .isPresent()
+                                    || ExceptionUtils.findThrowableWithMessage(
+                                                    err,
+                                                    "Either the environment variable AWS_WEB_IDENTITY_TOKEN_FILE or the javaproperty aws.webIdentityTokenFile must be set.")
+                                            .isPresent(),
```
Looks for messages is a bit fragile. Updating the AWS SDK could break this, and I doubt we would catch it in the test. Is there another way we can classify this?

dannycranmer replied:
Bump

dannycranmer replied:
Bump

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -56,6 +62,61 @@
+                            new KinesisDataStreamsException(
+                                    "Encountered non-recoverable exception relating to not being able to find the specified resources",
+                                    err));
+    private static final RetryValidationStrategy SDK_CLIENT_MISCONFIGURED_STRATEGY =
+            new RetryValidationStrategy(
+                    err -> ExceptionUtils.findThrowable(err, SdkClientException.class).isPresent(),
+                    err ->
+                            new KinesisDataStreamsException(
+                                    "Encountered non-recoverable exception relating to mis-configured client",
+                                    err));
```
This seems like it should be a class, that can be reused by Firehose sink, and other AWS sinks

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -56,6 +62,61 @@
+                            new KinesisDataStreamsException(
+                                    "Encountered non-recoverable exception relating to missing credentials",
+                                    err));
+    private static final RetryValidationStrategy NON_RECOVERABLE_EXCEPTION_STRATEGY =
+            new RetryValidationStrategy(
+                    err -> ExceptionUtils.findThrowable(err, Error.class).isPresent(),
+                    err ->
+                            new KinesisDataStreamsException(
+                                    "Encountered non-recoverable exception in the Kinesis Data Streams Sink",
+                                    err));
```
Same as above, can this pull pulled to module it can be reused from?

dannycranmer commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkRetryValidationStrategies.java:
```diff
@@ -0,0 +1,39 @@
+                    err ->
+                            ExceptionUtils.findThrowable(err, InterruptedException.class)
+                                    .isPresent(),
+                    err -> new FlinkException("Running job was cancelled"));
```
> Running job was cancelled

Are we confident enough to make this claim here? I do not think a Validation Strategy should know about Flink jobs. We can just more generally say "Thread was interrupted".

dannycranmer commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/RetryValidationStrategy.java:
```diff
@@ -0,0 +1,60 @@
+
+/** Class acting as a classifier for exception by failed requests to be retried by sink. */
+@Internal
+public class RetryValidationStrategy {
```
The word `Validation` in this class is confusing. Seems like it is a strategy determining when to retry validation. Suggest we make it more general `RetryStrategy` or maybe `ExceptionClassifier` but make sure there are no clashes with existing classes. This class is in the `sink` package but it looks more general than sink

vahmed-hamdy replied:
I tried going with `RetryStrategy` but this clashes with another interface in `flink-core`. `ExceptionClassifier` seems suitable.
I also agree with having it in a more general module, `flink-core` I believe should be suitable.

dannycranmer commented on flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/RetryValidationStrategyTest.java:
```diff
@@ -0,0 +1,119 @@
+        try {
+            int failedCalculation = 100 / 0;
+        } catch (Exception e) {
+            ARITHMETIC_EXCEPTION_STRATEGY.shouldRetry(e, caughtExceptionReference::set);
```
nit: Why not just instantiate and pass the exception in?

dannycranmer commented on flink-connectors/flink-connector-kinesis/pom.xml:
```diff
@@ -316,6 +316,7 @@ under the License.
 							<promoteTransitiveDependencies>true</promoteTransitiveDependencies>
 							<artifactSet combine.children="append">
 								<includes>
+									<include>org.apache.flink:flink-core:*</include>
```
Why shade flink-core? We do not want to do this, it is provided

dannycranmer commented on flink-connectors/flink-sql-connector-aws-kinesis-data-streams/pom.xml:
```diff
@@ -55,6 +55,7 @@
 						<configuration>
 							<artifactSet>
 								<includes>
+									<include>org.apache.flink:flink-core</include>
```
Why shade flink-core? We do not want to do this, it is provided

dannycranmer commented on flink-core/src/main/java/org/apache/flink/util/ThrowableWrapper.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    @Override
+    public ThrowableWrapper clone() {
+        ThrowableWrapper clonedSuppressor = new ThrowableWrapper(validator, throwableMapper);
+        Optional.ofNullable(this.chainedWrapper)
+                .ifPresent(wrapper -> clonedSuppressor.chainedWrapper = wrapper.clone());
+        return clonedSuppressor;
+    }
```
Why do we need a deep clone here? I see you using this to create the chains, but the chain will overwrite the `chainedWrapper` reference

dannycranmer commented on flink-core/src/main/java/org/apache/flink/util/ThrowableWrapper.java:
```diff
@@ -0,0 +1,81 @@
+
+    public static ThrowableWrapper build(ThrowableWrapper... classifiers) {
+        for (int i = 1; i < classifiers.length; ++i) {
+            classifiers[i - 1].chainedWrapper = classifiers[i];
```
There is a chance that chained wrappers can form a loop here. Unlikely, but possible (if `classifiers[length-1].chainedWrapper == classifiers[0]`)

dannycranmer commented on flink-core/src/main/java/org/apache/flink/util/ThrowableWrapper.java:
```diff
@@ -0,0 +1,81 @@
+                err -> ExceptionUtils.findThrowableWithMessage(err, message).isPresent(), mapper);
+    }
+
+    public static ThrowableWrapper build(ThrowableWrapper... classifiers) {
```
Semantically the name of this method is wrong. It does not build anything

dannycranmer commented on flink-core/src/main/java/org/apache/flink/util/ThrowableWrapper.java:
```diff
@@ -0,0 +1,81 @@
+
+/** Util class acting as a classifier for exception to suppress. */
+@Internal
+public class ThrowableWrapper {
```
I am not keen on the name `ThrowableWrapper`. It is a bit too generic to not be too meaningful. Given that it is in `flink-core` module it is hard to understand when to use this

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSCredentialExceptionWrappers.java:
```diff
@@ -0,0 +1,45 @@
+
+/** Class containing Exception Wrappers for AWS credenetial failures. */
+@Internal
+public class AWSCredentialExceptionWrappers {
```
Should this be `AWSCredentialRetryableExceptionClassifier` ?

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-firehose/src/test/java/org/apache/flink/connector/firehose/sink/KinesisFirehoseSinkITCase.java:
```diff
@@ -126,6 +123,70 @@ public void test() throws Exception {
+
+    private List<String> getSampleDataGenerated() throws JsonProcessingException {
+        List<String> expectedElements = new ArrayList<>();
+        for (int i = 1; i < NUMBER_OF_ELEMENTS; i++) {
```
This will add `NUMBER_OF_ELEMENTS - 1` to the list, is that expected?

dannycranmer commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/util/RetryableExceptionClassifier.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/** Util class acting as a classifier for exception to suppress. */
```
nit: I would like to see a better comment here, this is not technically a util class.

dannycranmer commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/util/RetryableExceptionClassifier.java:
```diff
@@ -0,0 +1,90 @@
+                err -> ExceptionUtils.findThrowable(err, type).isPresent(), mapper);
+    }
+
+    public static RetryableExceptionClassifier withRootCauseWithMessage(
```
I cannot see any usages of this, am I missing something or can we remove it?

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -56,6 +61,31 @@
+
+    private static final RetryableExceptionClassifier KINESIS_RETRY_VALIDATION_STRATEGY =
+            RetryableExceptionClassifier.createChain(
+                    GENERAL_ERROR_STRATEGY.clone(),
```
Cloning a singleton instance is an odd way to achieve this. Factories with builder methods or Builders are a more standard approach. For instance `AsyncSinkThrowableWrappers.generalErrorStrategy()`, can we please refactor this?

dannycranmer commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/util/RetryableExceptionClassifier.java:
```diff
@@ -0,0 +1,90 @@
+        importedWrappers.add(taleClassifier);
+        for (int i = 1; i < classifiers.length; ++i) {
+            if (importedWrappers.contains(classifiers[i])) {
+                continue;
```
I think we should throw an `InvalidArgumentException` here

dannycranmer reviewed (approved):
Thanks @vahmed-hamdy, LGTM, will merge after CI passes

vahmed-hamdy commented:
@flinkbot azure run

vahmed-hamdy commented:
@flinkbot run azure

vahmed-hamdy commented:
@flinkbot run azure

vahmed-hamdy commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

vahmed-hamdy commented:
@flinkbot run azure

vahmed-hamdy commented:
@flinkbot run azure

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-firehose/src/test/java/org/apache/flink/connector/firehose/sink/KinesisFirehoseSinkITCase.java:
```diff
@@ -78,42 +76,36 @@
-        s3AsyncClient = createS3Client(mockFirehoseContainer.getEndpoint(), httpClient);
-        firehoseAsyncClient = getFirehoseClient(mockFirehoseContainer.getEndpoint(), httpClient);
-        iamAsyncClient = createIamClient(mockFirehoseContainer.getEndpoint(), httpClient);
+        s3HttpClient = createHttpClient(mockFirehoseContainer.getEndpoint());
+        firehoseHttpClient = createHttpClient(mockFirehoseContainer.getEndpoint());
+        iamHttpClient = createHttpClient(mockFirehoseContainer.getEndpoint());
+        s3AsyncClient = createS3Client(mockFirehoseContainer.getEndpoint(), s3HttpClient);
+        firehoseAsyncClient =
+                getFirehoseClient(mockFirehoseContainer.getEndpoint(), firehoseHttpClient);
+        iamAsyncClient = createIamClient(mockFirehoseContainer.getEndpoint(), iamHttpClient);
+        env = StreamExecutionEnvironment.getExecutionEnvironment();
```
Why did you split out to separate HTTP clients?

CrynetLogistics replied:
So I can keep a reference to the HTTP clients to close

dannycranmer replied:
This was already done though. Before the AWS SDK clients were sharing a single HTTP client. I wondered if there is a reason we needed to use 3 instead of 1?

dannycranmer replied:
As discussed, this appears to address a classloader flaky test issue

dmvk replied:
Can you elaborate?

dannycranmer replied:
@CrynetLogistics reported that flaky classloader test was reproduced locally and this fixes it, can you please elaborate here as per our discussion?

CrynetLogistics replied:
Sharing a single http client between the 3 service clients causes the ci and local to fail intermittently. Frequency approx. 1 in 6. This is reproducible. Changing them to use separate ones fixes this issue and I have ran it over 380 times on my local without failure. The issue relates to `classloaders` being closed prior to usage of the clients.

dmvk replied:
Can you be more specific? This doesn't really explain what's causing the issue. What exect class-loader is being closed and how it relates to the AWS client? Can you point to a specific place in the code that causes the issue?

dmvk replied:
There is something sketchy going with thread locals and XML parser service loading. This can signal some deeper issues, so we should make sure why exactly this is happening, even though we have a "workaround" that works for some reason.

```
Feb 10 10:57:30 Caused by: java.lang.IllegalStateException: Trying to access closed classloader. Please check if you store classloaders directly or indirectly in static fields. If the stacktrace suggests that the leak occurs in a third party library and cannot be fixed immediately, you can disable this check with the configuration 'classloader.check-leaked-classloader'.
Feb 10 10:57:30 	at org.apache.flink.runtime.execution.librarycache.FlinkUserCodeClassLoaders$SafetyNetWrapperClassLoader.ensureInner(FlinkUserCodeClassLoaders.java:164)
Feb 10 10:57:30 	at org.apache.flink.runtime.execution.librarycache.FlinkUserCodeClassLoaders$SafetyNetWrapperClassLoader.getResources(FlinkUserCodeClassLoaders.java:188)
Feb 10 10:57:30 	at java.util.ServiceLoader$LazyIterator.hasNextService(ServiceLoader.java:348)
Feb 10 10:57:30 	at java.util.ServiceLoader$LazyIterator.hasNext(ServiceLoader.java:393)
Feb 10 10:57:30 	at java.util.ServiceLoader$1.hasNext(ServiceLoader.java:474)
Feb 10 10:57:30 	at javax.xml.stream.FactoryFinder$1.run(FactoryFinder.java:352)
Feb 10 10:57:30 	at java.security.AccessController.doPrivileged(Native Method)
Feb 10 10:57:30 	at javax.xml.stream.FactoryFinder.findServiceProvider(FactoryFinder.java:341)
Feb 10 10:57:30 	at javax.xml.stream.FactoryFinder.find(FactoryFinder.java:313)
Feb 10 10:57:30 	at javax.xml.stream.FactoryFinder.find(FactoryFinder.java:227)
Feb 10 10:57:30 	at javax.xml.stream.XMLInputFactory.newInstance(XMLInputFactory.java:154)
Feb 10 10:57:30 	at software.amazon.awssdk.protocols.query.unmarshall.XmlDomParser.createXmlInputFactory(XmlDomParser.java:124)
Feb 10 10:57:30 	at java.lang.ThreadLocal$SuppliedThreadLocal.initialValue(ThreadLocal.java:284)
Feb 10 10:57:30 	at java.lang.ThreadLocal.setInitialValue(ThreadLocal.java:180)
Feb 10 10:57:30 	at java.lang.ThreadLocal.get(ThreadLocal.java:170)
Feb 10 10:57:30 	at software.amazon.awssdk.protocols.query.unmarshall.XmlDomParser.parse(XmlDomParser.java:44)
Feb 10 10:57:30 	at software.amazon.awssdk.protocols.xml.internal.unmarshall.XmlResponseParserUtils.parse(XmlResponseParserUtils.java:56)
Feb 10 10:57:30 	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.parseResponse(AwsXmlPredicatedResponseHandler.java:114)
Feb 10 10:57:30 	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleResponse(AwsXmlPredicatedResponseHandler.java:95)
Feb 10 10:57:30 	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:84)
Feb 10 10:57:30 	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:42)
Feb 10 10:57:30 	at software.amazon.awssdk.core.internal.handler.BaseClientHandler.lambda$successTransformationResponseHandler$6(BaseClientHandler.java:232)
```

dannycranmer replied:
What is strange is why these clients are using the Flink user classloader to start with. This test is running an Embedded Mini Cluster, I assume that Flink is reusing the classloader from the unit test (a guess). If this is the case it could be a race condition between the Flink job finishing and and clients closing. 

We also discussed using a sync client, as the stack trace we saw previously came from one of the AWS SDK thread pools. But I would like to understand why this happens.

CrynetLogistics replied:
Any ideas on what might be the root cause here? This is something I spent Wednesday, Thursday and Friday looking at last week, with unfortunately no breakthrough. However, I did find that separate http clients resolved the issue.

Would you be able to assist? I believe the root cause is not in this codebase but in one of the sdk clients.

CrynetLogistics replied:
@dmvk I spent the past hour trying to reproduce it... I think I'm close, I have pushed what I have so far on to [FLINK-25846-commit-f2e288b](https://github.com/CrynetLogistics/flink/tree/FLINK-25846-commit-f2e288b). It hangs locally rather than failing with `classloader`.

The main issue is because we rebased this branch on master quite a few times now, something may have been introduced to change this behaviour. Anyhow it should be at or near this commit (f2e288b) in this PR's history.

CrynetLogistics replied:
@dmvk Just failed in this [PR](https://github.com/apache/flink/pull/18686) and it's corresponding [CI build](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=31136&view=results).

Since this is something that is currently impacting `master` and it is apparent that this PR fixes it, would you be happy to merge this and raise a follow up ticket for investigation?

CrynetLogistics replied:
But I know this would fix the problem, since I was able to reproduce it locally, and the final commit resolves the issue. Surely merging this is a better solution.

dmvk replied:
It fixes the problem and you don't know why, what kind of fix is that?

CrynetLogistics replied:
@dmvk Do I rebase on top of the commit you just pushed removing the `@Ignore` and merge this commit once the CI passes?

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-firehose/src/test/java/org/apache/flink/connector/firehose/sink/KinesisFirehoseSinkITCase.java:
```diff
@@ -78,42 +76,36 @@
     @After
     public void teardown() {
         System.clearProperty(SdkSystemSetting.CBOR_ENABLED.property());
-        AWSGeneralUtil.closeResources(
-                httpClient, s3AsyncClient, firehoseAsyncClient, iamAsyncClient);
+        s3HttpClient.close();
+        firehoseHttpClient.close();
+        iamHttpClient.close();
+        s3AsyncClient.close();
+        firehoseAsyncClient.close();
+        iamAsyncClient.close();
```
Why not use the util to close these?

CrynetLogistics replied:
I will do.

dannycranmer replied:
This was resolved but not updated?

dmvk replied:
Sorry, lack of time is not an excuse for not doing things properly...

dannycranmer replied:
Since we need to close the additional clients, please make the change

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-firehose/src/test/java/org/apache/flink/connector/firehose/sink/testutils/KinesisFirehoseTestUtils.java:
```diff
@@ -73,4 +81,22 @@ public static void createDeliveryStream(
+                .map(data -> mapper.writeValueAsString(ImmutableMap.of("data", data)));
+    }
+
+    public static List<String> getSampleDataGenerated(int endValue) throws JsonProcessingException {
```
nit: This method name does not make sense to me, should it be `getSampleData`/`generateSampleData`/`createSampleData`

CrynetLogistics replied:
4 hours ago intended to. I've reassessed and do not plan to implement due to lack of time.

CrynetLogistics replied:
done.

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-firehose/src/main/java/org/apache/flink/connector/firehose/sink/KinesisFirehoseSinkWriter.java:
```diff
@@ -65,6 +70,8 @@ private static SdkAsyncHttpClient createHttpClient(Properties firehoseClientProp
     private static FirehoseAsyncClient createFirehoseClient(
             Properties firehoseClientProperties, SdkAsyncHttpClient httpClient) {
+        AWSGeneralUtil.validateAwsConfiguration(firehoseClientProperties);
+        AWSGeneralUtil.validateWebIdentityTokenFileCredentialsProvider(firehoseClientProperties);
```
Why are we calling an explicit validation for WebIdentifyToken here? Why not a generic top level one? This does not scale if more and more identify providers are added. Firehose should not care explicitly about which one to validate.

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSGeneralUtil.java:
```diff
@@ -369,4 +367,35 @@ public static void closeResources(SdkAutoCloseable... resources) {
+            if (credentialProviderType.equals(WEB_IDENTITY_TOKEN)) {
+                getCredentialsProvider(config).resolveCredentials();
+            }
+        } catch (Throwable e) {
```
Can you catch something more specific that Throwable here? Why catch throwable? This is usually considered bad practise

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSGeneralUtil.java:
```diff
@@ -369,4 +367,35 @@ public static void closeResources(SdkAutoCloseable... resources) {
+            CredentialProvider credentialProviderType =
+                    getCredentialProviderType(config, AWSConfigConstants.AWS_CREDENTIALS_PROVIDER);
+            if (credentialProviderType.equals(WEB_IDENTITY_TOKEN)) {
+                getCredentialsProvider(config).resolveCredentials();
```
Where is this code executed? Is it on client or Task Manager? If client, then client will need AWS creds, this is not idea if job is being submitted from a host will limited permissions

vahmed-hamdy replied:
This is executed on the TM by the sink writer, the web Identity token falls back to env variables so verifying any missing credentials is not possible by verifying the configs only. I thought it would be more thorough to catch any failure on credential resolving and report it .
I believe the code is going to be re-executed by the client when sending requests using same credentials later, why should a failure here be ignored or not an indicator of bad credentials?

dannycranmer replied:
If this code runs on the TM it is ok. I just wanted to avoid the situation where the job fails to start because client has insufficient permissions

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure DO IT

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

CrynetLogistics commented:
@flinkbot run azure

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-firehose/src/test/java/org/apache/flink/connector/firehose/sink/KinesisFirehoseSinkITCase.java:
```diff
@@ -123,20 +115,21 @@ public void test() throws Exception {
-                listBucketObjects(
-                        createS3Client(mockFirehoseContainer.getEndpoint(), httpClient),
-                        BUCKET_NAME);
+        SdkAsyncHttpClient s3RetrievalHttpClient =
+                createHttpClient(mockFirehoseContainer.getEndpoint());
+        S3AsyncClient s3RetrievalAsyncClient =
+                createS3Client(mockFirehoseContainer.getEndpoint(), s3RetrievalHttpClient);
```
These clients are not closed

dmvk replied:
Again, these need to be closed...

CrynetLogistics commented:
All comments addressed. Thank you.

CrynetLogistics commented:
|| @dannycranmer  
|| Why did you split out to separate HTTP clients?

I will revert to the previous, where it is using a single http client. Please continue to consider this PR as the other commits  relate to necessary improvements to the sinks.

CrynetLogistics commented:
@flinkbot run azure

dmvk commented:
Here is the code I've used during debugging, it could provide some more context for understanding the issue better. https://gist.github.com/dmvk/6f3fbf5a6432becd076a9f7989ab8bdd

CrynetLogistics commented:
Thank you. I have understood the issue. I have now changed my usage of Netty to the Apache synchronous http client. I have also removed the `@Ignore` flag for the integration test case.

CrynetLogistics commented:
@flinkbot run azure

dmvk commented:
Also I would rather avoid switching the client implementations now, because that might require another investigation whether there is some kind of shared thread pool underneath (which might be a pattern in the SDK).

Only think we have to do is to construct a separate ELG group (of reasonable size) for the test suite and for each writer.

dannycranmer commented:
> Now the problem is that the Netty thread inherits a context classloader from the thread that has created it

I re-read the excellent deep dive, thanks @dmvk. From what I understand there would be no issue for Flink jobs submitted to a cluster, since they will use the FlinkUserClassloader. However, as we saw with the tests, when using the mini cluster this issue could resurface. Therefore I agree we should fix in `AWSGeneralUtil`.

dmvk commented:
One thing to note that there is a separate `FlinkUserClassloader` per-job (probably even per TM-slot, I'm not sure). So if there would be multiple jobs running on the same TM and one of them terminates (and closes it's classloader), we might IMO see this on a real cluster as well.

CrynetLogistics commented:
@dmvk Thank you for the in-depth view of the issue. I have created a [PR#18733](https://github.com/apache/flink/pull/18733) relating to this change.

vahmed-hamdy commented:
Since, @CrynetLogistics has isolated the Integration test fix in #18733 , 
I have rebased this PR now to address issues in [FLINK-25846] and [FLINK-25848] only.
@dannycranmer I am trying to find related comments but they seem to be addressed, could you please re-review?!

vahmed-hamdy commented:
@flinkbot run azure

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSCredentialRetryableExceptionClassifiers.java:
```diff
@@ -0,0 +1,47 @@
+                                err));
+    }
+
+    public static RetryableExceptionClassifier getSdkClientMisconfiguredExceptionClassifier() {
+        return RetryableExceptionClassifier.withRootCauseOfType(
+                SdkClientException.class,
+                err ->
+                        new AWSAuthenticationException(
+                                "Encountered non-recoverable exception relating to mis-configured client",
+                                err));
+    }
+}
```
It seems wrong to bucket all `SdkClientException` as `AWSAuthenticationException`. Why not just bubble the RC up here? I would rather not introduce incorrect exception wrappers

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSAuthenticationException.java:
```diff
@@ -0,0 +1,36 @@
+import org.apache.flink.annotation.Internal;
+
+/**
+ * A {@link IllegalStateException} wrapper indicating the exception was thrown from AWS credentials.
```
This comment does not make sense "thrown from AWS credentials"

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSAuthenticationException.java:
```diff
@@ -0,0 +1,36 @@
+ * A {@link IllegalStateException} wrapper indicating the exception was thrown from AWS credentials.
+ */
+@Internal
+public class AWSAuthenticationException extends IllegalStateException {
```
I do not think `IllegalStateException` semantically fits here:

> Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation.

https://docs.oracle.com/javase/7/docs/api/java/lang/IllegalStateException.html

We are throwing this when bad creds are provided right? So a better fit might be `IllegalArgumentException`? In any case, the purpose of this class is not really clear. If we are just wrapping any and all exceptions that arise from authentication then we should extend `RuntimeException`

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSGeneralUtil.java:
```diff
@@ -374,4 +371,34 @@ public static void closeResources(SdkAutoCloseable... resources) {
+        try {
+            validateAwsConfiguration(config);
+            getCredentialsProvider(config).resolveCredentials();
+        } catch (IllegalStateException | IllegalArgumentException | SdkClientException e) {
+            throw new AWSAuthenticationException(
+                    String.format(
+                            "Failed to create client using %s provider",
+                            getCredentialProviderType(
+                                    config, AWSConfigConstants.AWS_CREDENTIALS_PROVIDER)),
+                    e);
+        }
+    }
```
Again, how about if there are other issues thrown here that are not related to authentication? `SdkClientException`

>Direct Known Subclasses:
AbortedException, ApiCallAttemptTimeoutException, ApiCallTimeoutException, Crc32MismatchException, EndpointDiscoveryFailedException, ExpiredTokenException, NonRetryableException, RetryableException, SdkJsonGenerator.JsonGenerationException


Consider letting these bubble up, or wrapping more specific exceptions

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSGeneralUtil.java:
```diff
@@ -374,4 +371,34 @@ public static void closeResources(SdkAutoCloseable... resources) {
+        }
+    }
+
+    private static void validateCredentialProvider(Properties config) {
+        // value specified for AWSConfigConstants.AWS_CREDENTIALS_PROVIDER needs to be
+        // recognizable
+        try {
+            getCredentialsProvider(config);
+        } catch (IllegalArgumentException e) {
+            StringBuilder sb = new StringBuilder();
+            for (CredentialProvider type : CredentialProvider.values()) {
+                sb.append(type.toString()).append(", ");
+            }
+            throw new IllegalArgumentException(
+                    "Invalid AWS Credential Provider Type set in config. Valid values are: "
+                            + sb.toString());
+        }
+    }
```
There are no unit tests in this module for the changes in this PR

dannycranmer commented on flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSCredentialRetryableExceptionClassifiers.java:
```diff
@@ -0,0 +1,47 @@
+                                err));
+    }
+
+    public static RetryableExceptionClassifier getSdkClientMisconfiguredExceptionClassifier() {
```
Why does a misconfigured SDK client result in `AWSAuthenticationException` ?

vahmed-hamdy replied:
I agree it is an inaccurate mapping, 
As mentioned above I would prefer to remove the wrapper `AWSAuthenticationException` entirely and bubble up the root cause.

dannycranmer commented:
@flinkbot run azure

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/test/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkITCase.java:
```diff
@@ -310,7 +533,15 @@ public Scenario withPartitionKeyGenerator(
+        }
+
         private void prepareStream(String streamName) throws Exception {
+            if (streamName.equals("do-not-create-new-stream")) {
+                return;
+            }
```
nit: This seems like a hack? Can we update the Scenario to disable stream creation instead? You could skip if stream name is null for example? At least you could extract this String to a constant since it is heavily reused

dannycranmer commented on flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java:
```diff
@@ -58,6 +63,31 @@
+
+    private static final RetryableExceptionClassifier NON_RECOVERABLE_EXCEPTION_STRATEGY =
+            RetryableExceptionClassifier.withRootCauseOfType(
+                    Error.class,
```
I think generic `Error` should be handled by the core classifier. Classifiers created in concrete sink implementation should only care about destination specific errors, such as `ResourceNotFoundException` above

dannycranmer commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/util/RetryableExceptionClassifier.java:
```diff
@@ -0,0 +1,83 @@
+        this.chainedClassifier = null;
+    }
+
+    public boolean shouldSuppress(Throwable err, Consumer<Exception> throwableConsumer) {
```
Semantically this method is confusing. Since this is a `RetryableExceptionClassifier` would it be better to invert this to `shouldRetry`. Thoughts?

vahmed-hamdy replied:
I agree, we have been trying to align the name with a general function of exception classification and its current use case of validating retryable exceptions.
maybe `FatalExceptionClassifier` and `isFatal` would serve the purpose better.

dannycranmer commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/util/RetryableExceptionClassifier.java:
```diff
@@ -0,0 +1,83 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.base.sink.util;
```
Seems odd to put this in a `util` package since it is not a Utility class, and has a clear responsibility. How about moving this to `.throwable` similar to `org.apache.flink.runtime.throwable.ThrowableClassifier`

dannycranmer commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkRetryableExceptionClassifiers.java:
```diff
@@ -0,0 +1,38 @@
+public class AsyncSinkRetryableExceptionClassifiers {
+    public static RetryableExceptionClassifier getInterruptedExceptionClassifier() {
+        return RetryableExceptionClassifier.withRootCauseOfType(
+                InterruptedException.class, err -> new FlinkException("Thread was interrupted"));
```
We are not passing the root cause (`err`) in to the `FlinkException`

CrynetLogistics commented:
@flinkbot run azure
